packages/app/src/server/routes/api/cortex.ts

```diff
@@ -0,0 +1,143 @@
+/**
+ * Copyright 2020 Opstrace, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import httpProxy from "http-proxy";
+import express from "express";
+import querystring from "querystring";
+
+import { log } from "@opstrace/utils/lib/log";
+
+export const RING_TYPES = [
+  "ingester",
+  "compactor",
+  "store-gateway",
+  "ruler",
+  "alertmanager"
+] as const;
+
+type RingType = typeof RING_TYPES[number];
+
+export const cortexProxy = httpProxy.createProxyServer({ ignorePath: true });
+/**
+ * bodyParser middleware (which we use earlier in chain to parse POST body content), doesn't play
+ * well with the http-proxy, since the body is parsed and altered. It results in POSTs hanging until the connection times out.
+ *
+ * This is a workaround to restream the already parsed body, for the proxy target to consume.
+ */
+cortexProxy.on("proxyReq", function onProxyReq(proxyReq, req, res) {
+  // @ts-ignore
+  if (!req.body || !Object.keys(req.body).length) {
+    return;
+  }
+
+  const contentType = proxyReq.getHeader("Content-Type");
+  const writeBody = (bodyData: string) => {
+    proxyReq.setHeader("Content-Length", Buffer.byteLength(bodyData));
+    proxyReq.write(bodyData);
+  };
+
+  if (contentType === "application/json") {
+    // @ts-ignore
+    writeBody(JSON.stringify(req.body));
```

Comment:
This is wild :-) :fire:  Just to summarize, the HTTP proxy architecture we use here
- reads the request bytes, deserializes as JSON (this is what bodyParser is doing)
- reserializes to a byte sequence again and only then passes the request on to the proxy target

Instead, it should transparently forward-stream the body byte sequence.

I am optimistic that there is a better solution; using canonical tools from the NodeJS/express ecosystem. What about

https://github.com/chimurai/http-proxy-middleware which uses https://github.com/http-party/node-http-proxy under the hood (the same library used here via `import httpProxy from "http-proxy";`).

There is a comment in https://github.com/chimurai/http-proxy-middleware#intercept-and-manipulate-requests which suggests that this middleware should be applied before the bodyParser middleware:

> ... which is used to fix proxied POST requests when bodyParser is applied before this middleware.

This makes sense -- then the proxy middleware has access to the "raw" request body in a streaming fashion.

The same pointer, but way more explicit, is provided in 
https://www.npmjs.com/package/express-http-proxy#middleware-mixing

> Proxy requests and user responses are piped/streamed/chunked by default.

> If you use 'https://www.npmjs.com/package/body-parser' you should declare it AFTER the proxy configuration, otherwise original 'POST' body could be modified and not proxied correctly.

Reply:
cc @MatApple

Reply:
This is fine for a first pass, @jgehrcke and I talked about this offline. We can move the order of middleware and fix this problem - I'll take a look at this after we land this PR